docs/life-of-a-request.md

```diff
@@ -0,0 +1,200 @@
+or debug an existing one. Let's try to follow a request through the
+Zulip codebase, and dive deep into how each part works.
+
+## A request is sent to the server, and handled by Nginx
+
+Everything in Zulip goes through nginx, when deployed in production.
+For the most part you don't need to know how this works, except for when
+it isn't working. Nginx does the first level of routing--deciding which
+application will serve the request (or deciding to serve the request
+itself for static content).
+
+## Nginx secures traffic with SSL
+
+If you visit your zulip server in your browser and discover that your
+traffic isn't being properly encrypted, an nginx configuration is the
+likely culprit.
+
+## Static files are served directly by Nginx
+
+Static files include javascript, css, static assets (like emoji, avatars),
+and user uploads (if stored locally and not on S3).
+
+```
+/home/zulip/prod-static
+```
+
+## Nginx routes other requests between tornado and django
+
+All our connected clients hold open long-polling connections so that
+they can recieve events (messages, presence notifications, and so on) in
+real-time. Events are served by Zulip's `tornado` application.
+
+Nearly every other kind of request is served by the `zerver` Django
+application.
+
+# Django routes the request to a view in urls.py files
+
+There are various urls.py files throughout the server codebase, and you
+can read about it more in [the directory structure doc](http://zulip.readthedocs.io/en/latest/directory-structure.html).
+
+The main Zulip Django app is `zerver`. The routes are found in
+```
+zproject/urls.py
+zproject/legacy_urls.py
+```
+
+There are HTML-serving, REST API, legacy, and webhook url patterns.
+
+## Views serving HTML are internationalized by server path
+
+If you look in [zproject/urls.py](https://github.com/zulip/zulip/blob/master/zproject/urls.py), you'll see something called
+`i18n_urls`. These urls show up in the address bar of the browser, and
+serve HTML.
+
+For example, the `/hello` page (preview [here](https://zulip.com/hello/))
+gets translated in Chinese at `zh-cn/hello/` (preview [here](https://zulip.com/zh-cn/hello/)).
+
+Note the `zh-cn` prefix--that url pattern gets added by `i18n_patterns`.
+
+## API endpoints use REST
+
+With the exception of Webhooks (which we do not control), and legacy
+endpoints, and logged-out endpoints, Zulip uses REST for its API. This
+means that we use:
+
+* POST for creating something new where you don't have a unique ID
+* PUT for creating or completely replacing something for which you have
+  a unique ID
+* DELETE for deleting something
+* PATCH for updating or editing attributes of something.
+* GET to get something (read-only)
+* HEAD to check the existence of something to GET, without getting it;
+  useful to check a link without downloading a potentially large link
+* OPTIONS (handled automatically, see more below)
+* POST for everything else
+
```

Comment:
In the section above it might be worthwhile to choose between the form "VERB for verbing" ("POST for creating") and the form "VERB to verb" ("HEAD to check"). I prefer the latter, but I don't have a strong argument one way or the other.